File: 212__Word_Search_II.py
class Solution(object):

    def buildTrie(self, words):
        trie = {}
        for word in words:
            cur = trie
            for char in word:
                cur = cur.setdefault(char, {})
            cur["word"] = word
        return trie


    def dfs(self, board, i, j, trie):
        # found one
        if trie.get("word"):
            self.res.append(trie["word"])
            # de-duplicate
            trie["word"] = None
        if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]) or board[i][j] not in trie:
            return
        c, board[i][j] = board[i][j], "#"
        self.dfs(board, i - 1, j, trie[c])
        self.dfs(board, i + 1, j, trie[c])
        self.dfs(board, i, j - 1, trie[c])
        self.dfs(board, i, j + 1, trie[c])
        board[i][j] = c

    def findWords(self, board, words):
        """
        :type board: List[List[str]]
        :type words: List[str]
        :rtype: List[str]
        """
        # Backtracking + Trie
        trie = self.buildTrie(words)
        self.res = []
        for i in range(len(board)):
            for j in range(len(board[0])):
                self.dfs(board, i, j, trie)
        return self.res

    

File: Java/212__Word_Search_II.java
class Solution {
    public List<String> findWords(char[][] board, String[] words) {
        List<String> result = new ArrayList<>();
        int m = board.length, n = board[0].length;
        for (String word : words) {
            boolean check = false;
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    check = helper(board, word, 0, i, j);
                    if (check) break;
                }
                if (check) break;
            }
            if (check) result.add(word);
        }
        return result;
    }

    private boolean helper(char[][] board, String word, int index, int i, int j) {
        int m = board.length, n = board[0].length;
        if (index == word.length()) return true;
        if (i < 0 || i >= m || j < 0 || j >= n) return false;
        if (board[i][j] != word.charAt(index)) return false;
        char old_ij = board[i][j];
        board[i][j] = '*';
        boolean result = helper(board, word, index + 1, i + 1, j) || helper(board, word, index + 1, i - 1, j) || helper(board, word, index + 1, i, j + 1) || helper(board, word, index + 1, i, j - 1);
        board[i][j] = old_ij;
        return result;
    }
}
